packages/devextreme/js/__internal/data_controller.ts

```diff
@@ -0,0 +1,278 @@
+import { Deferred } from '@js/core/utils/deferred';
+
+import { extend } from '../core/utils/extend';
+import { isDefined } from '../core/utils/type';
+import ArrayStore from '../data/array_store';
+import { DataSource } from '../data/data_source/data_source';
+import { normalizeDataSourceOptions } from '../data/data_source/utils';
+
+interface DataSourceType {
+  _userData: unknown;
+  _pageSize: number;
+  dispose: () => void;
+  load: (loadOptions?: unknown) => Promise<unknown>;
+  loadSingle: (propName: string, propValue: unknown) => Promise<unknown>;
+  loadOptions: () => unknown;
+  cancel: (operationId: number) => void;
+  cancelAll: () => void;
+  filter: (filter?: unknown) => unknown;
+  group: (group?: unknown) => unknown;
+  items: () => unknown[];
+  key: () => string;
+  keyOf: (item: unknown) => string;
+  paginate: (paginate?: unknown) => unknown;
+  pageIndex: <TIdx = number | undefined>(pageIndex: TIdx) => TIdx extends number
+    ? undefined
+    : number;
+  reload: () => Promise<unknown>;
+  select: (...args: unknown[]) => unknown;
+  searchExpr: (searchExpr?: unknown) => unknown;
+  searchOperation: (searchOperation?: string) => unknown;
+  searchValue: (searchValue?: unknown) => unknown;
+  store: () => DataSourceType;
+  totalCount: () => number;
+  isLastPage: () => boolean;
+  isLoaded: () => boolean;
+  isLoading: () => boolean;
+  on: (eventName: string, eventHandler: () => void) => DataSourceType;
+  off: (eventName: string, eventHandler: () => void) => DataSourceType;
+  _addSearchFilter: (storeLoadOptions: unknown) => void;
+  _applyMapFunction: <TData = unknown>(data: TData) => TData;
+}
+
+interface DataControllerOptions extends Record<string, unknown> {
+  key: string;
+}
+
+class DataController {
+  private _isSharedDataSource = false;
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error
+  private _dataSource: DataSourceType;
+
+  private _keyExpr: string;
+
+  constructor(dataSourceOptions: unknown[] | DataSourceType, { key }: DataControllerOptions) {
+    this._keyExpr = key;
+    this.updateDataSource(dataSourceOptions);
+  }
+
+  _updateDataSource(dataSourceOptions: DataSourceType): void {
+    if (dataSourceOptions) {
```

Comment:
If we write this check this way we can get rid of an additional level of indent:
```javascript
if (!dataSourceOptions) {
    return;
}
```